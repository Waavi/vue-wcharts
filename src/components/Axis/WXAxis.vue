<template>
    <g :id="id">
        <line
            v-if="!hideLine"
            :x1="x1"
            :y1="y1"
            :x2="x2"
            :y2="y2"
            :stroke="axisStylesCmp.stroke"
        />
        <template
            v-for="(tick, index) in ticks"
        >
            <g
                :key="`tick-${index}`"
                :text-anchor="textAnchor"
            >
                <line
                    v-if="!hideTickMark"
                    :x1="tick.mark.x1"
                    :y1="tick.mark.y1"
                    :x2="tick.mark.x2"
                    :y2="tick.mark.y2"
                    :stroke="markStylesCmp.stroke"
                />
                <slot
                    name="tickText"
                    v-bind="tick.text"
                >
                    <TickText
                        v-bind="tick.text"
                        :styles="labelStylesCmp"
                    />
                </slot>
            </g>
        </template>
    </g>
</template>

<script>
import VueTypes from 'vue-types'
import axisMixin from '../../mixins/axis'

export default {
    name: 'WXAxis',
    axis: 'x',
    mixins: [axisMixin],
    props: {
        space: VueTypes.arrayOf(VueTypes.number).def([0, 20, 24, 20]),
        textOffsetY: VueTypes.number.def(20),
    },
}
</script>
