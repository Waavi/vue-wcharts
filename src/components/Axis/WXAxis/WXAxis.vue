<template>
    <g
        v-if="typeof tickY === 'number'"
        :id="id"
        class="WXAxis"
    >
        <line
            v-if="lineProps"
            v-bind="lineProps"
        />
        <g
            v-for="({ tickX, value, label, textProps }, index) in ticksWithInfo"
            :key="label"
            class="AxisTickAndMark"
            :transform="`translate(${tickX}, 0)`"
        >
            <line
                v-if="!hideTickMark && tickX !== undefined"
                class="AxisMark"
                :x1="0"
                :y1="tickY"
                :x2="0"
                :y2="isOnTop ? tickY + tickLength : tickY - tickLength"
                v-bind="actualStyles.tick"
            />
            <slot
                v-if="tickX !== undefined"
                name="tick"
                :value="value"
                :label="label"
                :index="index"
                :styles="actualStyles.tickLabel"
            >
                <WAxisText
                    v-if="label || label === 0"
                    class="AxisTick"
                    :x="0"
                    :y="tickY"
                    :value="label"
                    v-bind="textProps"
                    :styles="actualStyles.tickLabel"
                    #default="{ value }"
                >
                    <slot
                        name="tickText"
                        :value="value"
                    >
                        {{ String(value) }}
                    </slot>
                </WAxisText>
            </slot>
        </g>

        <!-- <slot
            name="label"
            v-bind="labels"
            :styles="actualStyles.label"
        >
            <WAxisLabel
                v-if="label"
                v-bind="labels"
                :styles="actualStyles.label"
            />
        </slot>

        <line
            v-if="!hideNegativeAxis && negativeAxis"
            v-bind="negativeAxis"
            :style="negativeAxisStylesCmp"
        /> -->
    </g>
</template>

<script>
import VueTypes from 'vue-types'
import outerElementMixin from '../../../mixins/outerElementMixin'
import { AXIS_TYPE, AXIS_TYPE_LIST, AXIS_DIMENSION } from '../axisMixin'
import printableAxisMixin from '../printableAxisMixin'
import WAxisText from '../WAxisText/WAxisText.vue'

const XAXIS_POSITION = {
    BOTTOM: 'bottom',
    TOP: 'top',
}

export default {
    name: 'WXAxis',
    dimension: AXIS_DIMENSION.X,
    components: {
        WAxisText,
    },
    mixins: [
        printableAxisMixin,
        outerElementMixin,
    ],
    props: {
        id: VueTypes.string.def(AXIS_DIMENSION.X),
        type: VueTypes.oneOf(AXIS_TYPE_LIST).def(AXIS_TYPE.NUMERIC),
        position: VueTypes.oneOf([XAXIS_POSITION.BOTTOM, XAXIS_POSITION.TOP]).def(XAXIS_POSITION.BOTTOM),
        height: VueTypes.number.def(30),
        offset: VueTypes.number.optional,

        labelAlign: VueTypes.oneOf(['start', 'middle', 'end']).def('end'),
        negativeAxisStyles: VueTypes.object.def({}),
    },
    computed: {
        layoutInOuterArea () {
            const {
                position, height, invisible, offset,
            } = this
            return {
                reference: 'canvas',
                order: -1, // it should be the first one
                position,
                left: 0,
                width: '100%',
                height: invisible || typeof offset === 'number' ? 0 : height,
            }
        },

        actualRange () {
            const { Chart: { canvas } } = this
            return [canvas.left, canvas.left + canvas.width]
        },

        isOnTop () {
            return this.position === XAXIS_POSITION.TOP
        },
        lineY () {
            const {
                Chart: { canvas }, outerLayout, isOnTop, offset,
            } = this
            if (typeof offset === 'number') {
                const actualOffset = (offset <= 1 ? offset * canvas.height : offset)
                return outerLayout.y + (isOnTop ? +1 : -1) * actualOffset
            }
            return isOnTop ? outerLayout.y + outerLayout.height : outerLayout.y
        },
        lineProps () {
            const {
                invisible, hideLine, lineY, outerLayout, actualStyles,
            } = this
            if (invisible || hideLine) return undefined
            return {
                x1: outerLayout.x,
                y1: lineY,
                x2: outerLayout.x + outerLayout.width,
                y2: lineY,
                style: actualStyles.line,
            }
        },
        tickY () {
            const { lineY, tickLength, isOnTop } = this
            return isOnTop ? lineY - tickLength : lineY + tickLength
        },
        ticksWithInfo () {
            const {
                actualTicks, tickLength, isOnTop,
            } = this
            const verticalPadding = tickLength + 2
            const textProps = {
                'dominant-baseline': isOnTop ? 'baseline' : 'hanging',
                'text-anchor': 'middle',
                dx: '1',
                dy: ((isOnTop ? -verticalPadding : verticalPadding)).toString(),
            }
            return actualTicks.map(({ value, scaledValue, label }) => ({
                value,
                tickX: scaledValue,
                label,
                textProps,
            }))
        },
    },
}
</script>

<style lang="scss" scoped>
.AxisTickAndMark {
    transition: all 1s;
}
</style>
