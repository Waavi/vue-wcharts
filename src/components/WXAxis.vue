<template>
    <g stroke="#999">
        <line
            :x1="x1"
            :y1="y1"
            :x2="x2"
            :y2="y2"
        />
    </g>
</template>

<script>
import VueTypes from 'vue-types'

export default {
    name: 'WXAxis',
    type: 'axis',
    inject: ['Cartesian'],
    props: {
        datakey: VueTypes.string.isRequired,
        space: VueTypes.arrayOf(VueTypes.number).def([0, 20, 24, 20]),
    },
    computed: {
        points () {
            return this.Cartesian.dataset.map(value => value.name)
        },
        x1 () {
            return this.Cartesian.canvas.x0
        },
        y1 () {
            return this.Cartesian.canvas.y1
        },
        x2 () {
            return this.Cartesian.canvas.x1
        },
        y2 () {
            return this.Cartesian.canvas.y1
        },
    },
}
</script>
