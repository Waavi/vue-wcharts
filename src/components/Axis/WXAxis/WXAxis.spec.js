import { shallowMount } from '@vue/test-utils'
import WXAxis from './WXAxis.vue'

describe('Common/WDot', () => {
    const dataset = [
        {
            name: 'Page A',
            one: 1000,
            two: 2400,
            three: 2400,
        },
        {
            name: 'Page B',
            one: 3000,
            two: 1398,
            three: 2210,
        },
        {
            name: 'Page C',
            one: 2000,
            two: 9800,
            three: 2290,
        },
    ]

    const canvas = {
        x0: 40,
        y0: 10,
        width: 540,
        height: 366,
        x1: 580,
        y1: 376,
    }

    const bounds = {
        max: 8800,
        min: 0,
    }

    const propsData = {
        datakey: 'name',
    }

    const provide = {
        Chart: {
            dataset,
            canvas,
            bounds,
            snap: {},
            space: [20, 20, 20, 20],
            padding: [20, 20, 20, 20],
            spaceObjects: [10, 10, 10, 10],
            yScale: a => a,
        },
    }

    const defaultConfig = {
        propsData,
        provide,
    }

    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WXAxis, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })
})
