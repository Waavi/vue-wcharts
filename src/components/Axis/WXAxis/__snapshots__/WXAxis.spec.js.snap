// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Axis/WXAxis Should be render correctly 1`] = `
<g id="WXAxis">
  <line x1="40" y1="376" x2="580" y2="376" stroke="#cccccc"></line>
  <g text-anchor="middle">
    <line x1="60" y1="376" x2="60" y2="381" stroke="#cccccc"></line>
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="60" y="396" dy="" index="0" value="Page A" styles="[object Object]"></WTickText-stub>
  </g>
  <g text-anchor="middle">
    <line x1="310" y1="376" x2="310" y2="381" stroke="#cccccc"></line>
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="310" y="396" dy="" index="1" value="Page B" styles="[object Object]"></WTickText-stub>
  </g>
  <g text-anchor="middle">
    <line x1="560" y1="376" x2="560" y2="381" stroke="#cccccc"></line>
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="560" y="396" dy="" index="2" value="Page C" styles="[object Object]"></WTickText-stub>
  </g> <svg width="100%" height="100%"><text x="580" y="401" text-anchor="end" style="font-size: 12; fill: #414141; font-weight: 800;">
      X Axis
    </text></svg>
  <!---->
</g>
`;

exports[`Axis/WXAxis Should be render correctly with negative values 1`] = `
<g id="WXAxis">
  <line x1="40" y1="376" x2="580" y2="376" stroke="#cccccc"></line>
  <g text-anchor="middle">
    <line x1="60" y1="376" x2="60" y2="381" stroke="#cccccc"></line>
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="60" y="396" dy="" index="0" value="Page A" styles="[object Object]"></WTickText-stub>
  </g>
  <g text-anchor="middle">
    <line x1="310" y1="376" x2="310" y2="381" stroke="#cccccc"></line>
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="310" y="396" dy="" index="1" value="Page B" styles="[object Object]"></WTickText-stub>
  </g>
  <g text-anchor="middle">
    <line x1="560" y1="376" x2="560" y2="381" stroke="#cccccc"></line>
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="560" y="396" dy="" index="2" value="Page C" styles="[object Object]"></WTickText-stub>
  </g> <svg width="100%" height="100%"><text x="580" y="401" text-anchor="end" style="font-size: 12; fill: #414141; font-weight: 800;">
      X Axis
    </text></svg>
  <line x1="40" y1="0" x2="580" y2="0" style="stroke: #999; stroke-width: 1; stroke-dasharray: 0;"></line>
</g>
`;

exports[`Axis/WXAxis Should be render correctly without line 1`] = `
<g id="WXAxis">
  <!---->
  <g text-anchor="middle">
    <line x1="60" y1="376" x2="60" y2="381" stroke="#cccccc"></line>
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="60" y="396" dy="" index="0" value="Page A" styles="[object Object]"></WTickText-stub>
  </g>
  <g text-anchor="middle">
    <line x1="310" y1="376" x2="310" y2="381" stroke="#cccccc"></line>
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="310" y="396" dy="" index="1" value="Page B" styles="[object Object]"></WTickText-stub>
  </g>
  <g text-anchor="middle">
    <line x1="560" y1="376" x2="560" y2="381" stroke="#cccccc"></line>
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="560" y="396" dy="" index="2" value="Page C" styles="[object Object]"></WTickText-stub>
  </g> <svg width="100%" height="100%"><text x="580" y="401" text-anchor="end" style="font-size: 12; fill: #414141; font-weight: 800;">
      X Axis
    </text></svg>
  <!---->
</g>
`;

exports[`Axis/WXAxis Should be render correctly without ticks lines 1`] = `
<g id="WXAxis">
  <line x1="40" y1="376" x2="580" y2="376" stroke="#cccccc"></line>
  <g text-anchor="middle">
    <!---->
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="60" y="396" dy="" index="0" value="Page A" styles="[object Object]"></WTickText-stub>
  </g>
  <g text-anchor="middle">
    <!---->
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="310" y="396" dy="" index="1" value="Page B" styles="[object Object]"></WTickText-stub>
  </g>
  <g text-anchor="middle">
    <!---->
    <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="560" y="396" dy="" index="2" value="Page C" styles="[object Object]"></WTickText-stub>
  </g> <svg width="100%" height="100%"><text x="580" y="401" text-anchor="end" style="font-size: 12; fill: #414141; font-weight: 800;">
      X Axis
    </text></svg>
  <!---->
</g>
`;
